clients/venice-producer/src/main/java/com/linkedin/venice/producer/AbstractVeniceProducer.java

```diff
@@ -0,0 +1,425 @@
+import com.linkedin.venice.schema.SchemaReader;
+import com.linkedin.venice.schema.writecompute.DerivedSchemaEntry;
+import com.linkedin.venice.serialization.avro.AvroProtocolDefinition;
+import com.linkedin.venice.serialization.avro.SchemaPresenceChecker;
+import com.linkedin.venice.serializer.FastSerializerDeserializerFactory;
+import com.linkedin.venice.serializer.RecordSerializer;
+import com.linkedin.venice.stats.ThreadPoolStats;
+import com.linkedin.venice.utils.PartitionUtils;
+import com.linkedin.venice.utils.Utils;
+import com.linkedin.venice.utils.VeniceProperties;
+import com.linkedin.venice.utils.concurrent.BlockingQueueType;
+import com.linkedin.venice.utils.concurrent.ThreadPoolFactory;
+import com.linkedin.venice.writer.VeniceWriter;
+import com.linkedin.venice.writer.VeniceWriterFactory;
+import com.linkedin.venice.writer.VeniceWriterOptions;
+import com.linkedin.venice.writer.update.UpdateBuilder;
+import com.linkedin.venice.writer.update.UpdateBuilderImpl;
+import io.tehuti.metrics.MetricsRepository;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Properties;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Consumer;
+import org.apache.avro.Schema;
+import org.apache.avro.generic.GenericContainer;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.commons.lang.StringUtils;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+
+/**
+ * A generic implementation of the {@link VeniceProducer} interface
+ *
+ * @see VeniceProducer
+ */
+public abstract class AbstractVeniceProducer<K, V> implements VeniceProducer<K, V> {
+  private static final Logger LOGGER = LogManager.getLogger(AbstractVeniceProducer.class);
+  private static final DurableWrite DURABLE_WRITE = new DurableWrite();
+
+  private VeniceProperties producerConfigs;
+  private boolean configured = false;
+  private boolean closed = false;
+  private VeniceProducerMetrics producerMetrics;
+
+  private SchemaReader schemaReader;
+  private ThreadPoolExecutor producerExecutor;
+  private VeniceWriter<byte[], byte[], byte[]> veniceWriter;
+
+  private RecordSerializer<Object> keySerializer;
+  private static final Schema STRING_SCHEMA = Schema.create(Schema.Type.STRING);
+  private static final Schema INT_SCHEMA = Schema.create(Schema.Type.INT);
+  private static final Schema LONG_SCHEMA = Schema.create(Schema.Type.LONG);
+  private static final Schema FLOAT_SCHEMA = Schema.create(Schema.Type.FLOAT);
+  private static final Schema DOUBLE_SCHEMA = Schema.create(Schema.Type.DOUBLE);
+  private static final Schema BYTES_SCHEMA = Schema.create(Schema.Type.BYTES);
+  private static final Schema BOOL_SCHEMA = Schema.create(Schema.Type.BOOLEAN);
+
+  protected void configure(
```

Comment:
can we validate the store for if its write enabled or not ?

Reply:
Sure. I can make that change on the router. I will raise a separate PR

Reply:
https://github.com/linkedin/venice/pull/422